### PR TITLE
Advancements with postgres

### PR DESCRIPTION
**Pull request by flovilmart**
Initial status:
```
1011 specs, 0 failures, 423 pending specs
```

Aug. 15
```
987 specs, 0 failures, 161 pending specs
```

### CODE REVIEW

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

File: spec/GridStoreAdapter.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ var FilesController = require('../src/Controllers/FilesController').default;
 
 
 // Small additional tests to improve overall coverage
-describe("GridStoreAdapter",() =>{
+describe_only_db('mongo')("GridStoreAdapter",() =>{
```
Comment (flovilmart) [inline spec/GridStoreAdapter.js]: only mongo as we don't start the DB when in PG mode

File: spec/InstallationsRouter.spec.js [hunk 2]
```diff
@@ -3,10 +3,9 @@ var Config = require('../src/Config');
 var rest = require('../src/rest');
 var InstallationsRouter = require('../src/Routers/InstallationsRouter').InstallationsRouter;
 
-var config = new Config('test');
-
-describe_only_db(['mongo'])('InstallationsRouter', () => {
+describe('InstallationsRouter', () => {
   it('uses find condition from request.body', (done) => {
+    var config = new Config('test');
```
Comment (flovilmart) [inline spec/InstallationsRouter.spec.js]: move the config so we have a proper isolation

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
```
1010 specs, 0 failures, 298 pending specs
```

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
@drew-gross feel free to review when you have a minute. I feel this adapter pretty messy... Want to knock down as many tests as possible and then maybe refactor / extract. What do you think?

File: spec/ParseAPI.spec.js [hunk 3]
```diff
@@ -137,7 +162,12 @@ describe('miscellaneous', function() {
     // Create a new server to try to recreate the unique indexes
     .then(reconfigureServer)
     .catch(error => {
-      expect(error.code).toEqual(Parse.Error.DUPLICATE_VALUE);
+      on_db('mongo', () => {
+        expect(error.code).toEqual(Parse.Error.DUPLICATE_VALUE);
+      });
+      on_db('postgres', () => {
+        expect(error.code).toEqual('23505');
```
Comment (drew-gross) [inline spec/ParseAPI.spec.js]: I think this test should still be excluded in PG if it's not returning the right error code in PG. I thought `on_db` was going to be reserved for unsupported operations (like PG has no equivalent of mongo's $pullAll)

File: spec/ParseGlobalConfig.spec.js [hunk 4]
```diff
@@ -4,18 +4,18 @@ var request = require('request');
 var Parse = require('parse/node').Parse;
 let Config = require('../src/Config');
 
-describe('a GlobalConfig', () => {
+describe_only_db('mongo')('a GlobalConfig', () => {
```
Comment (drew-gross) [inline spec/ParseGlobalConfig.spec.js]: Also should just be disabled for PG, PG should be able to support Config

File: spec/ParseObject.spec.js [hunk 5]
```diff
@@ -268,16 +268,28 @@ describe('Parse.Object testing', () => {
     });
   });
 
-  it_exclude_dbs(['postgres'])("can set null", function(done) {
+  it("can set null", function(done) {
```
Comment (drew-gross) [inline spec/ParseObject.spec.js]: Probably you should be able to set null on all things in PG. We had this discussion about validations and how non-nullability should be a validation, but I think for now, its best to aim for API incompatibilities only when truly unavoidable

File: spec/ParseRelation.spec.js [hunk 6]
```diff
@@ -296,7 +296,7 @@ describe('Parse.Relation testing', () => {
     });
   });
 
-  it_exclude_dbs(['postgres'])("query on pointer and relation fields with equal", (done) => {
+  it_exclude_dbs('postgres')("query on pointer and relation fields with equal", (done) => {
```
Comment (drew-gross) [inline spec/ParseRelation.spec.js]: this looks fishy

File: spec/ParseRelation.spec.js [hunk 7]
```diff
@@ -374,7 +374,7 @@ describe('Parse.Relation testing', () => {
     });
   });
 
-  it_exclude_dbs(['postgres'])("or queries on pointer and relation fields", (done) => {
+  it_exclude_dbs('postgres')("or queries on pointer and relation fields", (done) => {
```
Comment (drew-gross) [inline spec/ParseRelation.spec.js]: also this

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 8]
```diff
@@ -427,11 +708,45 @@ export class PostgresStorageAdapter {
     const qs = `SELECT count(*) FROM $1:name ${wherePattern}`;
     return this._client.one(qs, values, a => +a.count);
   }
+
+  performInitialization({ SchemaController }) {
```
Comment (drew-gross) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: I think this is too much access, the DB adapters shouldn't be allowed to see the entire SchemaController.

File: src/Controllers/SchemaController.js [hunk 9]
```diff
@@ -76,7 +76,7 @@ const defaultColumns = Object.freeze({
     "pushTime":     {type:'String'},
     "source":       {type:'String'}, // rest or webui
     "query":        {type:'String'}, // the stringified JSON query
-    "payload":      {type:'Object'}, // the JSON payload,
+    "payload":      {type:'String'}, // the stringified JSON payload,
```
Comment (drew-gross) [inline src/Controllers/SchemaController.js]: Will this interact poorly with parse.com?

**Comment by drew-gross** [comment]
Looks amazing to me. I have a few comments about test strategy that I think should be fixed. The code has become pretty complex, but I think the basic structure is mostly sound, so no need to fix any of that right away.

File: spec/ParseObject.spec.js (hunk 5 above)
Comment (flovilmart) [inline spec/ParseObject.spec.js]: We cannot create a column without a type, so setting null at first is impossible, we'd need to update the test to reflect that.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 8 above)
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Indeed.

File: src/Controllers/SchemaController.js (hunk 9 above)
Comment (flovilmart) [inline src/Controllers/SchemaController.js]: Not really, it's set as a string already in the pushStatusHandler: 
See: 
https://github.com/ParsePlatform/parse-server/blob/master/src/pushStatusHandler.js#L27
(https://github.com/ParsePlatform/parse-server/blob/master/src/pushStatusHandler.js#L41)

**Comment by facebook-github-bot-9** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@flovilmart updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2510/files/349f2bc3778468e898c8f0a6d9b5f2f01271463c..9fca376002f3e1608d03853cc1028405305e6d05)

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2510/files/8e1069c5afab7fe55b2a9e96ecc71df772facb5d..45f021e5dda0cc2a7a3d0c43554a8ea7462406ed)

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2510/files/45f021e5dda0cc2a7a3d0c43554a8ea7462406ed..76422e643a3f3bfafc79f20ed2ce21865b6b135c)

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

File: spec/ParseAPI.spec.js [hunk 10]
```diff
@@ -13,6 +13,31 @@ const deepcopy = require('deepcopy');
 
 const userSchema = SchemaController.convertSchemaToAdapterSchema({ className: '_User', fields: Object.assign({}, SchemaController.defaultColumns._Default, SchemaController.defaultColumns._User) });
 
+describe_only_db('mongo')('miscellaneous', () => {
```
Comment (flovilmart) [inline spec/ParseAPI.spec.js]: moved there as it tries to create a DB with mongo etc...

**Pull request merged**